copycat/runtime/src/main/java/org/apache/kafka/copycat/storage/KafkaOffsetBackingStore.java

```diff
@@ -0,0 +1,367 @@
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.copycat.errors.CopycatException;
+import org.apache.kafka.copycat.util.Callback;
+import org.apache.kafka.copycat.util.ConvertingFutureCallback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Implementation of OffsetBackingStore that uses a Kafka topic to store offset data.
+ */
+public class KafkaOffsetBackingStore implements OffsetBackingStore {
+    private static final Logger log = LoggerFactory.getLogger(KafkaOffsetBackingStore.class);
+
+    public final static String OFFSET_STORAGE_TOPIC_CONFIG = "offset.storage.topic";
+
+    private final static long CREATE_TOPIC_TIMEOUT_MS = 30000;
+
+    private Time time;
+    private Map<String, ?> configs;
+    private String topic;
+    private Consumer<byte[], byte[]> consumer;
+    private Producer<byte[], byte[]> producer;
+    private HashMap<ByteBuffer, ByteBuffer> data;
+
+    private Thread thread;
+    private boolean stopRequested;
+    private Queue<Callback<Void>> readLogEndOffsetCallbacks;
+
+    public KafkaOffsetBackingStore() {
+        this(new SystemTime());
+    }
+
+    public KafkaOffsetBackingStore(Time time) {
+        this.time = time;
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        this.configs = configs;
+        topic = (String) configs.get(OFFSET_STORAGE_TOPIC_CONFIG);
+        if (topic == null)
+            throw new CopycatException("Offset storage topic must be specified");
+
+        data = new HashMap<>();
+        stopRequested = false;
+        readLogEndOffsetCallbacks = new ArrayDeque<>();
+    }
+
+    @Override
+    public void start() {
+        log.info("Starting KafkaOffsetBackingStore with topic " + topic);
+
+        producer = createProducer();
+        consumer = createConsumer();
+        List<TopicPartition> partitions = new ArrayList<>();
+
+        // Until we have admin utilities we can use to check for the existence of this topic and create it if it is missing,
+        // we rely on topic auto-creation
+        List<PartitionInfo> partitionInfos = null;
+        long started = time.milliseconds();
+        while (partitionInfos == null && time.milliseconds() - started < CREATE_TOPIC_TIMEOUT_MS) {
+            partitionInfos = consumer.partitionsFor(topic);
+            Utils.sleep(Math.min(time.milliseconds() - started, 1000));
+        }
+        if (partitionInfos == null)
+            throw new CopycatException("Offset topic wasn't created within the allotted period");
+
+        for (PartitionInfo partition : partitionInfos)
+            partitions.add(new TopicPartition(partition.topic(), partition.partition()));
+        consumer.assign(partitions);
```

Comment:
I agree that this seems much nicer.

You said "simple consumer mode". So this is the New Consumer in simple consumer mode, correct?

Will the New Consumer  automatically handle rebalances due to leader failover (broker failure where the partition leader changes), even when in simple consumer mode? Because that was a downside to the previous Simple Consumer -- you had to handle leadership changes on your own.